jsonrpc: Wait for the socket to be ready to read.

brbzull0 opened the pull request:
We found out that sometimes we hit a busy loop here, so adding a timeout to make sure we do not hangup forever here.

bneradt reviewed (commented):
lgtm. I'll let a non-Yahoo reviewer chime in.

cmcfarlen commented on src/shared/rpc/IPCSocketClient.cc:
```diff
@@ -30,12 +30,12 @@
 #include <tscore/ink_assert.h>
 #include <tscore/ink_sock.h>
-
+#include <iostream>
```
This header is unused.

cmcfarlen commented on src/shared/rpc/IPCSocketClient.cc:
```diff
@@ -117,23 +117,36 @@ IPCSocketClient ::send(std::string_view data)
 {
   if (this->is_closed()) {
     // we had a failure.
     return {};
```
I know it's not part of the changes, but this will return `NO_ERROR`? But the comment indicates a failure.  Could `UNKNOWN` be a better return?

cmcfarlen commented on src/shared/rpc/IPCSocketClient.cc:
```diff
@@ -117,23 +117,36 @@ IPCSocketClient ::send(std::string_view data)
+        // done!
+        break;
+      }
+    } while (ret != 0 && attempts_left > 0);
```
Should this check for readStatus timeout and error and also stop the loop? If you don't check for timeout, this function will wait for timeout_ms * attempts (10 seconds by default).  What errors would poll return that we would want to retry?

cmcfarlen reviewed (changes requested):
I am not sure this is correct.  Should you call read_ready if there were some bytes read in previous iterations of the outer loop? Shouldn't a timeout from read_ready stop the inner loop?   It seems like this will always wait until EOF or until (timeout_ms * attempts) have elapsed. The logic is a bit confusing, so maybe some comments about what is happening will clear it up.

brbzull0 commented:
> I am not sure this is correct. Should you call read_ready if there were some bytes read in previous iterations of the outer loop? Shouldn't a timeout from read_ready stop the inner loop? It seems like this will always wait until EOF or until (timeout_ms * attempts) have elapsed. The logic is a bit confusing, so maybe some comments about what is happening will clear it up.

Hi. It was weird indeed 🤮 

I changed it a bit now, single loop, check for the data in the socket, if available we read, if not we wait a few times unless an error occurs. 
Should be simpler now.

side note:
I am planning to clean up this in a future PR, some of the `ReadError` makes no more sense since we changed the buffer approach few months back.

cmcfarlen commented on src/shared/rpc/IPCSocketClient.cc:
```diff
@@ -117,35 +117,52 @@ IPCSocketClient ::send(std::string_view data)
+    nread = ::read(_sock, buf, to_read);
+    if (nread > 0) {
+      bs.save(nread);
+      continue;
```
If this doesn't break here, and there are no more bytes available, it seems this loop will wait the max time before returning.   Shouldn't we go ahead and return these bytes and if the caller needs more it can call this again?

brbzull0 replied:
Sorry, didn't clarify this, yes It will wait till  we have the data, or exausted the max time/tries.
Returning some data won't work as it will be a badly formed message, so I rather wait and see if we get the whole message before giving up. I see no benefit on making it more complicated. 

very simple protocol, and no need to retry, if any error, just run the command again and that's it(traffic_ctl or traffic_top).

cmcfarlen replied:
The max time default is 10 seconds.  Is the socket guaranteed to close or something?  How does `read_all` know the message is complete?

brbzull0 replied:
if you have a message for instance, to fetch records, or metrics and you have more than 6k metrics, then `10s` seems rasonable as it could take a while to send all the strings over the pipe.
 
I would expect `read` return `EOF` once done, if it's done then `read_all` knows there is no more data to read(ofc).  If there is any error different than EAGAIN then this will just stop and fail.

if you think tehre is  something that I am missing, please let me know. I'll be happy to address.

cmcfarlen reviewed (changes requested):
Please clarify if this loop should always wait the max time for any data vs returning after reading any.

brbzull0 commented:
> Please clarify if this loop should always wait the max time for any data vs returning after reading any.

The idea is to wait till EO, or an error(ofc), returning any data could be useless as the message will be incomplete(without trying again for more data), this should be simple, if there is any error, then run `traffic_ctl` again and that should be it. 
I can add some comments to make this clear in the code.

cmcfarlen reviewed (approved):
Thank you!

brbzull0 commented:
> Thank you!

Thank you for taking the time and provide good feedback! appreciated!
